### PR TITLE
Implementation

### PR DESCRIPTION
**Pull request by ka1zen**
Luminator:
set a session with username and login and a session id, and expose a fetch method that take an AxiosRequestConfig and:
- return AxiosResponse when the server respond with a 200 status,
- throw an error if a status is not in the STATUS_CODE_FOR_RETRY
- retry if the status is in STATUS_CODE_FOR_RETRY and refresh sessionId:
    - if the server respond with a 200 status it returns AxiosResponse
    - if it reach the setted threshold it throw an error

### CODE REVIEW

File: __tests__/unit/index.ts [hunk 1]
```diff
@@ -1,19 +1,106 @@
...
+        mockAxios.mockRejectedValue(failWith(status));
+        const spy: jest.SpyInstance = jest.spyOn(agent, 'switchSessionId');
+
+        try {
+          await agent.fetch({
+            method: 'GET',
+            url: 'https://lumtest.com/myip.json',
+          });
+        } catch (e) {
+          expect(spy).toHaveBeenCalledTimes(20);
```
Comment (rimiti) [inline __tests__/unit/index.ts]: `e` must be strictly checked. (message + typing)

File: __tests__/unit/index.ts [hunk 2]
```diff
@@ -1,19 +1,106 @@
...
+    });
   });
 
-  it("Should not work", async () => {
-    try {
+  describe('Should call switchSessionId with [403, 429, 502, 503] status and if the max allowed call with given ip reached' +
+    ', and to be called 20 times', () => {
+
+    Luminator.STATUS_CODE_FOR_RETRY.forEach((status: number) => {
+      it(`Should call switchSessionId with ${status} status, and to be called 20 times`, async () => {
```
Comment (rimiti) [inline __tests__/unit/index.ts]: It's not really clean to forEach `it`.

File: __tests__/unit/index.ts [hunk 3]
```diff
@@ -1,19 +1,106 @@
...
+            url: 'https://lumtest.com/myip.json',
+          });
+        } catch (e) {
+          expect(spy).toHaveBeenCalledTimes(20);
+        }
+      });
+    });
 
-    } catch (e) {
+    it(`Should call switchSessionId two time if the max query threshold is called`, async () => {
```
Comment (rimiti) [inline __tests__/unit/index.ts]: Description must be more explicit.

File: __tests__/unit/index.ts [hunk 4]
```diff
@@ -1,19 +1,106 @@
...
+      const agent = new Luminator('USERNAME', 'password');
```
Comment (rimiti) [inline __tests__/unit/index.ts]: `const agent = new Luminator('USERNAME', 'password');` is duplicated twice. You could move it into the describe scope.

File: src/luminator.ts [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+  private readonly password: string;
+  private readonly superProxy: string;
+  private readonly country: string;
+  private readonly port: number;
+
+  private superProxyHost: string;
+  private superProxyUrl: Proxy;
+  private sessionId: number;
+
+  constructor(username: string, password: string, superProxy: string = 'NL', country: string = 'fr', port: number = 22225) {
```
Comment (rimiti) [inline src/luminator.ts]: There are too many parameters. Could you use an object ?

File: src/luminator.ts [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+    this.username = username;
+    this.password = password;
+    this.superProxy = superProxy;
+    this.country = country;
+    this.port = port;
+    this.switchSessionId();
+  }
+
+  private static statusCodeRequiresExitNodeSwitch(statusCode: number): boolean {
```
Comment (rimiti) [inline src/luminator.ts]: - JSdoc is missing
- method name to renamed (must be more shorter & more explicit)

File: src/luminator.ts [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+    return [403, 429, 502, 503].includes(statusCode);
+  }
+
+  private static getRandomNumber(): number {
```
Comment (rimiti) [inline src/luminator.ts]: JSdoc is missing

Comment (rimiti) [reply src/luminator.ts]: - Method calle once, you could removed it.

File: src/luminator.ts [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+ * luminator doc
+ */
+class Luminator {
+  public static STATUS_CODE_FOR_RETRY: number[] = [403, 429, 502, 503];
+  private static readonly USER_AGENT: string =
+    'Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/41.0.2228.0 Safari/537.36';
+  private static readonly SWITCH_IP_EVERY_N_REQ: number = 30;
+  private static readonly MAX_FAILURES: number = 3;
+  private static readonly REQ_TIMEOUT: number = 60 * 1000;
+  private static readonly MAX_FAILURES_REQ: number = 11;
```
Comment (Simobain) [inline src/luminator.ts]: All those constants have been tested ? @ka1zen

File: src/luminator.ts [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+  private static statusCodeRequiresExitNodeSwitch(statusCode: number): boolean {
+    return [403, 429, 502, 503].includes(statusCode);
+  }
+
+  private static getRandomNumber(): number {
+    return Math.random();
+  }
+
+  private static getSessionId(): number {
+    return Math.trunc(Luminator.getRandomNumber() * 1000000);
```
Comment (rimiti) [inline src/luminator.ts]: `return Math.trunc(Math.random() * 1000000);`

File: src/luminator.ts [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+
+  private static statusCodeRequiresExitNodeSwitch(statusCode: number): boolean {
+    return [403, 429, 502, 503].includes(statusCode);
+  }
+
+  private static getRandomNumber(): number {
+    return Math.random();
+  }
+
+  private static getSessionId(): number {
```
Comment (rimiti) [inline src/luminator.ts]: JSdoc is missing

File: src/luminator.ts [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+    return Math.trunc(Luminator.getRandomNumber() * 1000000);
+  }
+
+  public async fetch(params: AxiosRequestConfig): Promise<AxiosResponse> {
```
Comment (rimiti) [inline src/luminator.ts]: JSdoc is missing

File: src/luminator.ts [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+    if (this.failuresCountReq >= Luminator.MAX_FAILURES_REQ) {
```
Comment (rimiti) [inline src/luminator.ts]: `failuresCountRequest` and `MAX_FAILURES_REQUEST`

File: src/luminator.ts [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+      throw new Error('MAX_FAILURES_REQ threshold reached');
+    } // break with too much failure
```
Comment (rimiti) [inline src/luminator.ts]: Comment not useful because you have a throw error above.

File: src/luminator.ts [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+    if (this.nReqForExitNode >= Luminator.SWITCH_IP_EVERY_N_REQ) {
```
Comment (rimiti) [inline src/luminator.ts]: nReqForExitNode  and SWITCH_IP_EVERY_N_REQ must be more explicit.

File: src/luminator.ts [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+
+    if (!this.haveGoodSuperProxy()) {
+      await this.switchSuperProxy();
+    }
+
+    let response: AxiosResponse;
+    try {
+      response = await axios(this.getAxiosRequestConfig(params));
+      this.onSuccess();
+    } catch (err) {
```
Comment (rimiti) [inline src/luminator.ts]: `err` must be typed.

File: src/luminator.ts [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+      this.failuresCountReq += 1;
```
Comment (rimiti) [inline src/luminator.ts]: failuresCountReq to renamed in failuresCountRequests

File: src/luminator.ts [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+      if (! Luminator.statusCodeRequiresExitNodeSwitch(err.status)) { // this could be 404 or other website error
```
Comment (rimiti) [inline src/luminator.ts]: - whitespace to remove
- Comment to remove ?

File: src/luminator.ts [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+        this.nReqForExitNode += 1;
+        throw err;
+      }
+      this.switchSessionId();
+      this.failCount += 1;
+    }
+
+    return response ? response : this.fetch(params);
```
Comment (rimiti) [inline src/luminator.ts]: This is not really beautiful but you might simplify that like:

return response || this.fetch(params);

File: src/luminator.ts [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+  }
+
+  public switchSessionId(): void {
```
Comment (rimiti) [inline src/luminator.ts]: jsdoc is missing

File: src/luminator.ts [hunk 20]
```diff
@@ -0,0 +1,160 @@
...
+    this.sessionId = Luminator.getSessionId();
+    this.nReqForExitNode = 0;
+    this.updateSuperProxyUrl();
+  }
+
+  private onSuccess():void {
```
Comment (rimiti) [inline src/luminator.ts]: - jsdoc is missing
- whutespace is missing `onSuccess(): void {`

File: src/luminator.ts [hunk 21]
```diff
@@ -0,0 +1,160 @@
...
+    this.failCount = 0;
+    this.nReqForExitNode += 1;
+    this.failuresCountReq = 0;
+  }
+
+  private getProxyOptions(): IProxyManagerOption {
```
Comment (rimiti) [inline src/luminator.ts]: jsdoc missing

File: src/luminator.ts [hunk 22]
```diff
@@ -0,0 +1,160 @@
...
+    return {
+      host: this.superProxyUrl.host,
+      port: this.superProxyUrl.port,
+      auth: `${this.superProxyUrl.auth.username}:${this.superProxyUrl.auth.password}`,
+    }
+  }
+
+  private getAxiosRequestConfig(params: AxiosRequestConfig): AxiosRequestConfig {
```
Comment (rimiti) [inline src/luminator.ts]: jsdoc is missing

File: src/luminator.ts [hunk 23]
```diff
@@ -0,0 +1,160 @@
...
+    return {
+      timeout: Luminator.REQ_TIMEOUT,
+      headers: { 'User-Agent': Luminator.USER_AGENT },
+      httpsAgent: new HttpsProxyAgent(this.getProxyOptions()),
+      ...params,
+    }
+  }
+
+  private haveGoodSuperProxy(): boolean {
```
Comment (rimiti) [inline src/luminator.ts]: jsdoc is missing

File: src/luminator.ts [hunk 24]
```diff
@@ -0,0 +1,160 @@
...
+    return this.superProxyHost && this.failCount < Luminator.MAX_FAILURES;
+  }
+
+  private getUsername(): string {
```
Comment (rimiti) [inline src/luminator.ts]: same

File: src/luminator.ts [hunk 25]
```diff
@@ -0,0 +1,160 @@
...
+    return `${this.username}-country-${this.country}-session-${this.sessionId}`;
+  }
+
+  private updateSuperProxyUrl(): void {
```
Comment (rimiti) [inline src/luminator.ts]: jsdoc

File: src/luminator.ts [hunk 26]
```diff
@@ -0,0 +1,160 @@
...
+      host: this.superProxyHost,
+      port: this.port,
+      auth: {
+        username: this.getUsername(),
+        password: this.password,
+      },
+    };
+  }
+
+  private getSuperProxyHost(): Promise<dns.LookupAddress> {
```
Comment (rimiti) [inline src/luminator.ts]: jsdoc

File: src/luminator.ts [hunk 27]
```diff
@@ -0,0 +1,160 @@
...
+    const address: dns.LookupAddress = await this.getSuperProxyHost();
+    this.superProxyHost = address.address;
+    this.updateSuperProxyUrl();
+  }
+}
+
+export {
+  Luminator,
+  IProxyManagerOption,
+  Proxy
```
Comment (rimiti) [inline src/luminator.ts]: A comma is missing

File: src/luminator.ts (hunk 8 above)
Comment (ka1zen) [reply src/luminator.ts]: Yes.

File: src/luminator.ts [hunk 28]
```diff
@@ -0,0 +1,160 @@
...
+    try {
+      response = await axios(this.getAxiosRequestConfig(params));
+      this.onSuccess();
+    } catch (err) {
+      this.failuresCountReq += 1;
+      if (! Luminator.statusCodeRequiresExitNodeSwitch(err.status)) { // this could be 404 or other website error
+        this.nReqForExitNode += 1;
+        throw err;
+      }
+      this.switchSessionId();
```
Comment (Simobain) [inline src/luminator.ts]: We switch the sessionId everytime we have an error code ?
No try again with the same session ?

**Comment by Simobain** [comment]
We don't have a pool of IP at the beginning ? Is it better to crawl website ? @ka1zen

**Comment by rimiti** [comment]
- The README.md should have an "Example" block with some use cases.

File: src/luminator.ts (hunk 28 above)
Comment (ka1zen) [reply src/luminator.ts]: if we have an error with one of the explicitly declared erros status code then we switchSession if the 429 was a warning it is better for all the community to switch the session (to keep the ip white listed)

File: __tests__/unit/index.ts (hunk 2 above)
Comment (ka1zen) [reply __tests__/unit/index.ts]: it's the best way to avoid repetition of the same code

File: src/luminator.ts [hunk 29]
```diff
@@ -0,0 +1,202 @@
...
+    password: string;
+  },
+}
+
+interface IProxyManagerOption {
+  host: string;
+  port: number;
+  auth: string;
+}
+// 'NL' 'fr' 22225
```
Comment (rimiti) [inline src/luminator.ts]: Comment to remove.

**Comment by rimiti** [comment]
LGTM

**Pull request merged**